sig-scalability/blogs/k8s-services-scalability-issues.md

```diff
@@ -0,0 +1,123 @@
+
+Now if you consider watches in the picture (there's one from each kube-proxy), the situation becomes even worse as the O(N^2) size traffic gets multiplied further with no. of watches (usually equal to #nodes in the cluster). Overall, this is a serious performance drawback affecting multiple components in the control-plane (apiserver, etcd, endpoints-controller, kube-proxy). The current endpoints API makes it hard to solve this problem without introducing breaking changes.
+
+#### Relevant issues/links
+
+- [#47787](https://github.com/kubernetes/kubernetes/issues/47787)
+- [#24552](https://github.com/kubernetes/kubernetes/issues/24552)
+- [#24553](https://github.com/kubernetes/kubernetes/issues/24553)
+- [#8190](https://github.com/kubernetes/kubernetes/issues/8190)
+
+#### Possible Solution(s)
+
+- Redesign endpoints API. Some properties we may like to have for the new API:
+  - Lives in an API group
+  - Not coupled to service resource lifecycle
+  - Can track both pod and non-pod endpoints
+  - Sharded into multiple etcd key/value pairs? (potentially in its own etcd)
+  - Has a Ready property
+  - Stable sharding & ordering
+  - Port ranges and other features we’ve put off for a long time
+
+
+## Define/measure performance SLIs for endpoints propagation
+
+#### Issue
+
+We need to measure the e2e endpoints propagation latency, i.e the time from when a 
+service is created until its endpoints are populated, given that all of the associated pods 
+are already running and ready. We also need to come up with reasonable SLOs for the 
+same and verify that they’re satisfied at scale.
+
+#### Relevant issues/links
+
+- [#10436](https://github.com/kubernetes/kubernetes/issues/10436)
+- This [doc on networking SLIs](https://docs.google.com/document/d/1kudI8uILO13ySkpE5HgNHvIGn-yXVQatpyM3YTVVUis) proposes one way of formulating the SLI
+
+#### Possible Solution(s)
+
+- Measure the SLI defined in the above doc and come up with a reasonable SLO
+
+
+## Master election shouldn’t consume unnecessary watch bandwidth
+
+#### Issue
+
+Currently we store the leader lock for master components like the scheduler in the corresponding endpoints object. This has the undesirable side effect of sending a notification down the kube-proxy <-> master (and kube-dns <-> master) watch every second, which is not actually required.
+
+#### Relevant issues/links
+
+- [#34627](https://github.com/kubernetes/kubernetes/issues/34627)
+
+#### Possible Solutions(s)
+
+- Avoid overloading endpoints as a lock API. We instead want to move to using the [new Lease API](https://github.com/kubernetes/kubernetes/pull/64246) (ETA 1.12). One potential challenge here is performing upgrades of HA clusters that are using the old endpoints-based locking approach to this new API approach without breaking them.
+- Another option is using configmaps, which is a supported option already in our client-go leader election library. But we don't choose this in favour of the above better option.
```

Comment:
Let's remove this - with Lease api that is designed for that particular purpose, this doesn't make sense.

Reply:
Done.